src/flash.rs

```diff
@@ -2,6 +2,303 @@
 
 use crate::pac::{flash, FLASH};
 
+pub const FLASH_START: u32 = 0x0800_0000;
+pub const FLASH_END: u32 = 0x080F_FFFF;
+
+const _RDPRT_KEY: u16 = 0x00A5;
+const KEY1: u32 = 0x45670123;
+const KEY2: u32 = 0xCDEF89AB;
+
+pub const SZ_1K: u16 = 1024;
+
+pub type Result<T> = core::result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum Error {
+    AddressLargerThanFlash,
+    AddressMisaligned,
+    LengthNotMultiple2,
+    LengthTooLong,
+    EraseError,
+    ProgrammingError,
+    WriteError,
+    VerifyError,
+    UnlockError,
+    LockError,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum SectorSize {
+    Sz1K = 1,
+    Sz2K = 2,
+    Sz4K = 4,
+}
+impl SectorSize {
+    const fn kbytes(self) -> u16 {
+        SZ_1K * self as u16
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub enum FlashSize {
+    Sz16K = 16,
+    Sz32K = 32,
+    Sz64K = 64,
+    Sz128K = 128,
+    Sz256K = 256,
+    Sz384K = 384,
+    Sz512K = 512,
+    Sz768K = 768,
+    Sz1M = 1024,
+}
+impl FlashSize {
+    const fn kbytes(self) -> u32 {
+        SZ_1K as u32 * self as u32
+    }
+}
+
+pub struct FlashReaderGuard<'a> {
+    _flash: &'a FlashWriter<'a>,
+    data: &'a [u8],
+}
```

Comment:
```suggestion
pub struct FlashReaderGuard<'b, 'a: 'b> {
    _flash: &'b FlashWriter<'a>,
    data: &'b [u8],
}
```

Reply:
So, no longer relevant because we'll be removing the FlashReaderGuard entirely, but I'm curious what is the purpose of creating this second lifetime specifier? I looked up the syntax of `'a: 'b` means "`'a` must last at least as long as `'b`", but what does that give us?

Reply:
The `'a` in `FlashWriter<'a>` is the lifetime of the borrow of `Parts`, we wouldn't need the `FlashReaderGuard` to borrow the `FlashWriter` for this same lifetime, so having another lifetime parameter helps with that, not sure if that would create real implications or the compiler would be smart enough to solve it.  Either way, it would be better because it wouldn't over constrain our implementation, but anyway, we probably won't need `FlashReaderGuard`...

Reply:
I see, that makes sense. Thanks!